File: Data Structures/02_Linked_Lists/14_inserting_node_into_a_sorted_dlist.cpp
#include <iostream>

using namespace std;

struct Node
{
    int data;
    Node *next;
    Node *prev;
};

int len = 0;
Node* SortedInsert(Node *head, int data);

void print_list(Node *head);

int main()
{
    int a[] = {1, 2, 5, 3, 1, 2, 3, 2};
    int size_a = sizeof(a) / sizeof(a[0]);

    Node *head = nullptr;

    for (int i = 0; i < size_a; i++) {
        head = SortedInsert(head, a[i]);
    }
    cout << "List: ";
    print_list(head);
    cout << "len: " << len << endl;

    return 0;
}

Node* SortedInsert(Node *head, int data)
{
    // Create new node with data
    Node *node = new Node();
    node->prev = node->next = nullptr;
    node->data = data;

    // If head is NULL or nullptr then the list is empty
    // then the only node in the list is the one you just created
    if (!head)
        return node;

    else {
        // For iteration purpose
        Node *current = head;

        // Since the list will be in ascending order, traverse the list
        // till there is node with data smaller than given value
        while(current->next && current->data < data)
            current = current->next;

        if (current->data > data) {

            // You may need to insert a node in between two nodes
            if (current->prev) {
                Node *prev = current->prev;

                prev->next = node;
                node->prev = prev;

                node->next = current;
                current->prev = node;
            }
            // or before the first node
            else {
                current->prev = node;
                node->next = current;
                head = node;
            }
        }
        else {
            // Insert in between two existing nodes
            if (current->next) {
                Node *next = current->next;

                next->prev = node;
                node->next = next;

                current->next = node;
                node->prev = current;
            }
            // or, insert after the last
            else {
                current->next = node;
                node->prev = current;
            }
        }
    }
    return head;
}

void print_list(Node *head)
{
    Node *temp = head;
    while(temp) {
        cout << temp->data << " ";
        temp = temp->next;
        len++;
    }
    cout << endl;
}

File: Data Structures/02_Linked_Lists/14_inserting_node_into_a_sorted_dlist.py
class Node(object):
    
    def __init__(self, data=None, next_node=None, prev_node=None):
        self.data = data
        self.prev = prev_node
        self.next = next_node

def SortedInsert(head, data):
    node = Node(data)

    if not head:
        return node
    else:
        current = head
        
        while current.next and current.data < data:
            current = current.next
        
        if current.data > data:
            if current.prev:
                prev = current.prev

                node.next = current
                node.prev = prev

                prev.next = node
                current.prev = node
            else:
                current.prev = node
                node.next = current
                head = node
        else:
            if current.next:
                next = current.next

                node.next = next
                node.prev = current

                next.prev = node
                current.next = node
            else:
                current.next = node
                node.prev = current
    return head

def print_list(head):
    temp = head
    while temp:
        print(temp.data, end=' ')
        temp = temp.next
    print()

head = None
data_set = [1, 9, 7, 3, 6, 8, 4]

for i in data_set:
    head = SortedInsert(head, i)

print_list(head)
